### PR TITLE
refactor: rewrite deno test, add Deno.test()

### PR DESCRIPTION
**Pull request by bartlomieju**
- rewrite `deno test` subcommand - no longer requires to download files from internet
- add `cli/testing.ts` - this file is a trimmed down copy of `std/testing/mod.ts` - has only `test()` and `runTests()` method
- Rust test `js_unit_test` now uses `Deno.test()` to create all test cases instead of relying on `std/testing/mod.ts` (this is just some dog-fooding)

`deno test` **does not** support globs expansion and exclusion. One needs to pass a full list of files that should be run - ie. you can still call `deno test */*_test.ts` but it is shell's job to expand the glob. 
It's really an MVP to take a step back and look at test module to provide essential things to write test. I expect that third party testing frameworks will arise (or be ported from node). So by no means this PR tries to establish a "batteries-included"  testing framework.

### CODE REVIEW

File: cli/js/lib.deno.ns.d.ts [hunk 1]
```diff
@@ -10,6 +10,31 @@ declare namespace Deno {
...
+    skip?: RegExp;
+    disableLog?: boolean;
+  }
+
+  export function runTests(opts: RunTestsOptions): Promise<void>;
+
+  export function runIfMain(
+    meta: ImportMeta,
+    opts?: RunTestsOptions
+  ): Promise<void>;
```
Comment (nayeemrmn) [inline cli/js/lib.deno.ns.d.ts]: Remove its usages in `cli/js/net_test.ts` and `cli/js/tls_test.ts`, replace its usage in `cli/js/unit_tests.ts` with an explicit `import.meta.main` check. Then you can get rid of this.

File: cli/test_runner.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    .filter(|p| {
+      let result = glob_set.is_match(p);
+      dbg!(p, result.clone());
+      result
+    })
+    .map(|p| {
+      dbg!("mapping");
+      Url::from_file_path(p).unwrap()
+    })
+    .collect()
```
Comment (nayeemrmn) [inline cli/test_runner.rs]: Isn't this https://github.com/denoland/deno_std/issues/576 again? What's wrong with the `glob` crate used for the formatter?

Comment (bartlomieju) [reply cli/test_runner.rs]:
```
Comparison with the glob crate
Supports alternate "or" globs, e.g., *.{foo,bar}.
Can match non-UTF-8 file paths correctly.
Supports matching multiple globs at once.
Doesn't provide a recursive directory iterator of matching file paths, although I believe this crate should grow one eventually.
Supports case insensitive and require-literal-separator match options, but doesn't support the require-literal-leading-dot option.
```
Our default test glob is: `**/?(*_)test.{js,ts}`. `glob` in `fmt` will be replaced with `globset`.

Comment (nayeemrmn) [reply cli/test_runner.rs]: Honestly that glob I wrote uses slightly inconsistent syntax. The proper [extended globbing](https://www.linuxjournal.com/content/bash-extended-globbing) syntax for OR is `@(a|b)`.

Does the `glob` crate support extended globbing? In other words would it support
`**/?(*_)test.@(js|ts)`

Comment (bartlomieju) [reply cli/test_runner.rs]: > Honestly that glob I wrote uses slightly inconsistent syntax. The proper [extended globbing](https://www.linuxjournal.com/content/bash-extended-globbing) syntax for OR is `@(a|b)`.
> 
> Does the `glob` crate support extended globbing? In other words would it support
> `**/?(*_)test.@(js|ts)`

I don't think so. I'm trying to port behavior of current test runner but `glob` in `std/` is whole other beast than the ones available in Rust - it requires a log of plumbing to be written

Comment (nayeemrmn) [reply cli/test_runner.rs]: Hmm, I see it doesn't. It doesn't to do much at all...

Comment (bartlomieju) [reply cli/test_runner.rs]: Yeah... `globset` is not really easier either; I can't really get it to work
```
[cli/test_runner.rs:30] g.glob() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner/**/?(*_)test.{js,ts}"
[cli/test_runner.rs:30] g.regex() = "(?-u)^/Users/biwanczuk/dev/deno/cli/tests/test_runner(?:/|/.*/).\\(.*_\\)test\\.(ts|js)$"
[cli/test_runner.rs:39] p.path() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner"
[cli/test_runner.rs:39] result.clone() = false
[cli/test_runner.rs:39] p.path() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner/foo.ts"
[cli/test_runner.rs:39] result.clone() = false
[cli/test_runner.rs:39] p.path() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner/test.js"
[cli/test_runner.rs:39] result.clone() = false
[cli/test_runner.rs:39] p.path() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner/bar_test.js"
[cli/test_runner.rs:39] result.clone() = false
[cli/test_runner.rs:39] p.path() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner/subdir"
[cli/test_runner.rs:39] result.clone() = false
[cli/test_runner.rs:39] p.path() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner/subdir/foo.ts"
[cli/test_runner.rs:39] result.clone() = false
[cli/test_runner.rs:39] p.path() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner/subdir/test.js"
[cli/test_runner.rs:39] result.clone() = false
[cli/test_runner.rs:39] p.path() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner/subdir/bar_test.js"
```

Comment (nayeemrmn) [reply cli/test_runner.rs]: With the `globset`/walking method, you can just costlessly expand the glob into `["**/test.js", "**/test.ts", "**/*_test.js", "**/*_test.ts"]` like it used to be.

Comment (bartlomieju) [reply cli/test_runner.rs]: Right, now it works :))

**Comment by nayeemrmn** [comment]
Once the `std` test suite is fully converted to use `Deno.test()` and `std/testing/runner.ts` is removed, can we make `runTests()` internal?

**Comment by bartlomieju** [comment]
> Once the `std` test suite is fully converted to use `Deno.test()` and `std/testing/runner.ts` is removed, can we make `runTests()` internal?

Still a long way to that... That said, having removed `runIfMain` leaves only `runTests` as a way to trigger tests and I see no reason to disallow users to write their own test script that does some fancy stuff.

**Comment by nayeemrmn** [comment]
> write their own test script that does some fancy stuff

Makes sense. If that's the purpose, though, we should make the exposed `Deno.runTests()` return result data and not do its own logging. That would make the current one unstable if not hidden...

**Comment by bartlomieju** [comment]
> > write their own test script that does some fancy stuff
> 
> Makes sense. If that's the purpose, though, we should make the exposed `Deno.runTests()` return result data and not do its own logging. That would make the current one unstable if not hidden...

Sure, but let's discuss after we land this one. 

CI still fails on glob expansion on Windows and I don't have a platform to test on 😕 I'm thinking about removing glob functionality completely for now

**Comment by nayeemrmn** [comment]
Yeah, I never realised how good globrex was. There's nothing that comes close in cargo.

For future: Do we have a way of shelling out small bits of work like this to JS? If not, we should. Then maybe we could bring globrex and the glob expander into `cli/js` 😆 at least until there is something in rustland.

File: cli/flags.rs [hunk 3]
```diff
@@ -22,15 +22,6 @@ macro_rules! sset {
...
 }
 
-macro_rules! std_url {
-  ($x:expr) => {
-    concat!("https://deno.land/std@v0.29.0/", $x)
-  };
-}
-
-/// Used for `deno test...` subcommand
-const TEST_RUNNER_URL: &str = std_url!("testing/runner.ts");
```
Comment (ry) [inline cli/flags.rs]: 🎉

**Comment by bartlomieju** [comment]
> Have you checked that if you insert a failing test into somewhere in cli/js that it will indeed still trigger failure? Same for std?
> 
> We need to be careful that we don’t silently ignore test failures when changing the runner.

Certainly, both cases throw as expected. Keep in mind that this PR does not change testing in `std/` - it only affects `cli/js/`. For `std/` PR take a look at #3930

File: cli/lib.rs [hunk 4]
```diff
@@ -408,6 +409,23 @@ async fn fmt_command(files: Option<Vec<String>>, check: bool) {
...
+  quiet: bool,
+) {
+  if let Some(test_file_path) =
+    test_runner::run_test_modules(include, fail_fast, quiet)
+  {
+    let mut flags = flags.clone();
+    flags
+      .argv
+      .push(test_file_path.to_string_lossy().to_string());
+    run_script(flags, test_file_path.to_string_lossy().to_string()).await;
```
Comment (nayeemrmn) [inline cli/lib.rs]: Shouldn't this be in `test_runner::run_test_modules()`? Also is `.deno.test.ts` being deleted?

**Comment by bartlomieju** [comment]
Great, once we land, I'll try to get PR with glob handling working

**Comment by ry** [comment]
I think as long as it took a directory it would be fine. I don't think that needs glob - you just walk the tree.

**Pull request merged**